src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -413,90 +417,130 @@ internal static void WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<Re
+        // 3. not bigger than 2^32 - 1 in total
+        // This function is also responsible for pinning the buffers if they
+        // are suitable and they must be unpinned after the I/O operation completes.
+        // It also returns a pointer with the segments to be passed to the
+        // Windows API, and the total size of the buffers that is needed as well.
+        // The pinned MemoryHandles and the pointer to the segments must be cleaned-up
+        // with the CleanupScatterGatherBuffers method.
+        private static unsafe bool TryPrepareScatterGatherBuffers<T, THandler>(IReadOnlyList<T> buffers,
+            THandler handler, out MemoryHandle[] handlesToDispose, out IntPtr segmentsPtr, out int totalBytes)
+            where THandler: struct, IMemoryHandler<T>
         {
-            if (!handle.IsAsync)
-            {
-                return ScheduleSyncReadScatterAtOffsetAsync(handle, buffers, fileOffset, cancellationToken);
-            }
+            int pageSize = s_cachedPageSize;
+            Debug.Assert(BitOperations.IsPow2(pageSize), "Page size is not a power of two.");
+            // We take advantage of the fact that the page size is
+            // a power of two to avoid an expensive modulo operation.
+            long alignedAtPageSizeMask = pageSize - 1;
+
+            int buffersCount = buffers.Count;
+            handlesToDispose = new MemoryHandle[buffersCount];
+            segmentsPtr = IntPtr.Zero;
+            totalBytes = 0;
+
+            // "The array must contain enough elements to store nNumberOfBytesToWrite bytes of data, and one element for the terminating NULL. "
+            long* segments = (long*) NativeMemory.Alloc((nuint)(buffersCount + 1), sizeof(long));
+            segments[buffersCount] = 0;
 
-            if (CanUseScatterGatherWindowsAPIs(handle))
+            bool success = false;
+            try
             {
-                long totalBytes = 0;
-                int buffersCount = buffers.Count;
+                long totalBytes64 = 0;
                 for (int i = 0; i < buffersCount; i++)
                 {
-                    totalBytes += buffers[i].Length;
+                    T buffer = buffers[i];
+                    int length = handler.GetLength(in buffer);
+                    totalBytes64 += length;
+                    if (length != pageSize || totalBytes64 > int.MaxValue)
+                    {
+                        return false;
+                    }
+
+                    MemoryHandle handle = handlesToDispose[i] = handler.Pin(in buffer);
+                    long ptr = segments[i] = (long)handle.Pointer;
+                    if ((ptr & alignedAtPageSizeMask) != 0)
+                    {
+                        return false;
+                    }
                 }
 
-                if (totalBytes <= int.MaxValue) // the ReadFileScatter API uses int, not long
+                segmentsPtr = (IntPtr)segments;
```

Comment:
Looks like all uses of this require casting it back to `long*`.  Why not just keep it as `long*` from the start, rather than casting away the type?

Reply:
`async` methods cannot be unsafe.

Reply:
Ok